backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/ColumnarCachedBatchSerializer.scala

```diff
@@ -0,0 +1,311 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.glutenproject.backendsapi.velox
+
+import io.glutenproject.GlutenConfig
+import io.glutenproject.backendsapi.BackendsApiManager
+import io.glutenproject.columnarbatch.ColumnarBatches
+import io.glutenproject.exec.ExecutionCtxs
+import io.glutenproject.execution.RowToVeloxColumnarExec
+import io.glutenproject.memory.arrowalloc.ArrowBufferAllocators
+import io.glutenproject.memory.nmm.NativeMemoryManagers
+import io.glutenproject.utils.{FieldRecordBatch, VectorLoaderWithField, VectorUnloaderWithField}
+import io.glutenproject.utils.ArrowRecordBatchType.{FieldBuffers, FieldNumRows, FiledNullCounts}
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.{InternalRow, SQLConfHelper}
+import org.apache.spark.sql.catalyst.expressions.{Attribute, Expression}
+import org.apache.spark.sql.columnar.{CachedBatch, CachedBatchSerializer}
+import org.apache.spark.sql.execution.VeloxColumnarToRowExec
+import org.apache.spark.sql.execution.columnar.DefaultCachedBatchSerializer
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.{StructField, StructType}
+import org.apache.spark.sql.utils.SparkArrowUtil
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.storage.StorageLevel
+import org.apache.spark.util.TaskResources
+
+import org.apache.arrow.vector.VectorSchemaRoot
+
+case class CachedColumnarBatch(
+    override val numRows: Int,
+    override val sizeInBytes: Long,
+    fieldBuffersArr: Array[FieldBuffers],
+    fieldNullCountsArr: Array[FiledNullCounts],
+    fieldNumRowsArr: Array[FieldNumRows])
+  extends CachedBatch {}
+
+// spotless:off
+/**
+ * Note that, Spark does not support cache a offheap object, and it converts `UnsafeRow` to
+ * `Array[Byte]` then cache it.
+ * So this serializer also use that way, by converting velox `ColumnarBatch` to
+ * `Array[Byte]` then cache it.
+ *
+ * There is a issue that, if we allocate a small onheap memory, then it may be cause the
+ * cached block spill to disk frequently.
+ *
+ * Feature:
+ * 1. This serializer supports column pruning without deserialize the `Array[Byte]`
+ * 2. TODO: support push down filter
+ * 3. Super TODO: cache offheap object directly if Spark framework support in the future
+ *
+ * The data transformation pipeline:
+ *
+ *   - Serializer ColumnarBatch -> CachedColumnarBatch
+ *     -> Export to Arrow VectorSchemaRoot
+ *     -> Unload to CachedColumnarBatch (onheap)
+ *
+ *   - Deserializer CachedColumnarBatch -> ColumnarBatch
+ *     -> Load to Arrow VectorSchemaRoot
```

Comment:
Since we are planning to remove the Arrow dependency in Gluten, can we use Velox's [deserialization interface](https://github.com/oap-project/velox/blob/175870db403ba9159d588fc772a5c936c27d0910/velox/row/UnsafeRowDeserializers.h#L903) to construct a RowVector instead of relying on the Arrow VectorSchemaRoot to call Arrow API? @ulysses-you @zhztheplayer @zhouyuan

Reply:
I agree with you, but it's not the trivial things, this pr also uses the row to columnar which depends on arrow.. I think we can improve these plan step by step in future prs. The first thing is to make row to columnar decouple with arrow.

Reply:
> I agree with you, but it's not the trivial things, this pr also uses the row to columnar which depends on arrow.. I think we can improve these plan step by step in future prs. The first thing is to make row to columnar decouple with arrow.

Let's try eliminating some of the Arrow usages in this patch. I just helped adding some code @ulysses-you

Reply:
After second thought, I think it's too hurried to decide to remove the Arrow dependency in Gluten. How to export/import schema ? How to manage the onheap memory ? If we can not clean up all the usage of Arrow in a short time, I think it's overkill to remove the usage of Arrow.

Reply:
agree, the key idea is to remove arrow from the I/O flow firstly, e.g., using velox native column vectors, thus the code logic will be simpler. the schema/metadata part indeed needs more refactor work - should be done in another epic 

thanks, -yuan